### PR TITLE
Add some BanManager improvements

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
This will be required for my next PR which actually implements banning. I wanted to split it up into two to make it more digestible.

### CODE REVIEW

File: source/agora/common/BanManager.d [hunk 1]
```diff
@@ -86,7 +86,7 @@ public class BanManager
         if (status.fail_count >= this.config.max_failed_requests)
         {
             status.fail_count = 0;
-            status.banned_until = this.getCurTime() + this.config.ban_duration;
+            this.banFor(address, this.config.ban_duration);
```
Comment (Geod24) [inline source/agora/common/BanManager.d]: Why not `ban` ?

File: source/agora/common/BanManager.d [hunk 2]
```diff
@@ -210,8 +210,9 @@ public class BanManager
 
     ***************************************************************************/
 
-    private Status* get (Address address)
+    private Status* get (Address address) @trusted nothrow pure
     {
+        scope(failure) assert(0);  // it will never throw
```
Comment (Geod24) [inline source/agora/common/BanManager.d:215]: `require` is not `nothrow` ??

Comment (AndrejMitrovic) [reply source/agora/common/BanManager.d:215]: It creates an item if it doesn't exist. The naming of `require` is very counter-intuitive.

File: source/agora/common/BanManager.d (hunk 1 above)
Comment (AndrejMitrovic) [reply source/agora/common/BanManager.d]: Good eye.

**Review by Geod24 — APPROVED** [review]
Side note: What happens if `ban` or `banFor` is called on an existing banned node ? Does `banFor` takes the max of both bans ?

File: source/agora/common/BanManager.d (hunk 1 above)
Comment (Geod24) [reply source/agora/common/BanManager.d]: Ping

**Comment by AndrejMitrovic** [comment]
Updated.

> Side note: What happens if ban or banFor is called on an existing banned node ? Does banFor takes the max of both bans ?

Ah very good question.. Should it add to the existing ban time, or use the max? Currently it just overrides whatever was set.

**Comment by AndrejMitrovic** [comment]
Maybe I should simplify this:

```
public void banForever() /// e.g. known bad nodes
private void ban()  // called internally in onFailedRequest
```

**Comment by Geod24** [comment]
No I don't think we should ever ban forever. IPs are dynamic.
As for the ban time, a conservative approach would be to implicitly extend and explicitly reduce.
In other word, if you ban an already ban client, unless you do it via a specially-designed functions, you only ever extend.
And at the moment I don't see a reason to extend / reduce bans ?

**Review by Geod24 — APPROVED** [review]

**Pull request merged**